test/NuGet.Core.Tests/NuGet.Common.Test/SynchronizationTest.cs

```diff
@@ -20,6 +20,15 @@ public class SynchronizationTests
 
         private readonly int DefaultTimeOut = (int)TimeSpan.FromMinutes(5).TotalMilliseconds;
 
+        public SynchronizationTests()
+        {
+            // Starting with .NET 7, we can delete lock files also on non-Windows platforms.
+            if (Environment.Version.Major >= 7)
+            {
+                Environment.SetEnvironmentVariable("NUGET_ConcurrencyUtils_DeleteOnClose", "1");
```

Comment:
I don't understand the intention/benefit of this.

There are no tests to validate the file either exists when the environment variable is not set, or that the file is deleted when it is set. Therefore, this doesn't change test results in any way, it's just extra code. If the intention is to not have the empty files left over after running tests, [NuGet.Common's `FileUtility` uses `ConcurrencyUtilities`](https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L68), and [there's a test for it](https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/test/NuGet.Core.Tests/NuGet.Common.Test/FileUtilityTests.cs#L107), meaning depending on the order the tests run, there either will or won't be a file left over from that test. I'm not 100% sure, but I think xunit runs tests in a non-deterministic order to try to help catch bugs that accidentally have static variable state dependencies.

Additionally, anyone who can successfully run our build scripts, [we run all test assemblies in the same process](https://github.com/NuGet/NuGet.Client/blob/ad81306fe7ada265cf44afb2a60a31fbfca978a2/build/build.proj#L413), so again, different behaviour depending on (non-deterministic?) order that tests and assemblies are executed.

Reply:
The envvar allows the test, or users, to opt-in to the delete behavior.
What would you like to change?

Reply:
In my opinion, all code is technical debt, we just have to choose what tech debt to accept. For example, code implementing features, and tests validating features work. For test code specifically, if this code doesn't change test results or catch regressions, then it should be removed.

I put myself in the shoes of someone who is reading this code for the first time and doesn't have the context of this PR, or the history behind the change. Maybe they're modifying `ConcurrencyUtilities` to add a new feature, or just trying to investigate some bug unrelated to `DeleteOnClose`. This code might make them stop and think "how does this affect the test? Is it safe to remove? Is it mandatory to keep?", and therefore slow down whatever work they're doing.  The easier code is to read, the easier it is for first-timers to understand, and for code reviewers to review.

If the intention is to test and validate that it works, then we need a complex test that uses `System.Diagnostics.Process` to launch new processes, like we have with our `dotnet.integration.tests` or `msbuild.integration.tests`, or `NuGet.CommandLine.FuncTests`. `Process` and `StartInfo` allow us to control the new process' environment variables, and then test both the "on" and "off" conditions, without risking flaky tests for tests that run in parallel, or before/after this test assembly runs when xunit runs multiple assemblies in the same process.

In short, I suggest removing this test file change, and not have any test changes in the PR. I don't feel like it adds anything.

Reply:
`ConcurrencyUtilities` uses [KeyedLock](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/KeyedLock.cs), which uses in-process `SemaphoreSlim` to more efficiently handle concurrency in process. So, it still wouldn't test the `DeleteOnClose` behaviour without the `System.Diagnostics.Process` idea to have multiple processes run in parallel.  Alternatively, the code needs to be refactored so the file locking can be tested independently from `KeyedLock`. In other words, the test would be a false positive as-is.